### DISCUSSION TITLE
Is block-spmm supported？

### DISCUSSION THREAD

**Discussion by mzy2240** [JuliaGPU/CUDA.jl#1737] [Q&A] [answered]
https://developer.nvidia.com/blog/accelerating-matrix-multiplication-with-block-sparse-format-and-nvidia-tensor-cores/

Aside cusparselt, just curious whether block-spmm is supported? Seems to ship with cusparse.

**Comment by maleadt** [comment, marked as answer]
Please don't use the issue tracker for questions.

---

In general, all library APIs are supported, the question is at what level:

- C wrappers: functions like `cusparseSpMM` can always be called directly, as CUDA.jl exposes the library's C APIs: https://github.com/JuliaGPU/CUDA.jl/blob/108e75f316052961124639af6c2b240bf12e8e80/lib/cusparse/libcusparse.jl#L6545-L6603
- Julia wrappers: to make functions more user-friendly (e.g., making them type-generic) we often provide slightly high level wrappers. In the case of `cusparseSpMM`, that's the `mm!` function: https://github.com/JuliaGPU/CUDA.jl/blob/108e75f316052961124639af6c2b240bf12e8e80/lib/cusparse/generic.jl#L177-L294
- interface integrations: at the highest level of abstraction, CUDA.jl tries to integrate with libraries like LinearAlgebra.jl. For `mm!` that's `LinearAlgebra.mul!`, which in certain cases should dispatch to `cusparseSpMM`: https://github.com/JuliaGPU/CUDA.jl/blob/7e58f28257c6597a19ef73c4b3fa06480e620235/lib/cusparse/interfaces.jl#L32-L103

What's required for *block* sparse gemm to work is the new BlockedEll storage format. That isn't supported by any of the high level wrappers, so you'd need to use the low-level `cusparseCreateBlockedEll` API, or provide your own integrations with the higher-level types (mimicking the support for CSR/CSC/COO/BSR storage types).
